spec/rubocop/cop/lint/out_of_range_regexp_ref_spec.rb

```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::Lint::OutOfRangeRegexpRef do
+  subject(:cop) { described_class.new(config) }
+
+  let(:config) { RuboCop::Config.new }
+
+  it 'registers an offense when out of range references are used for named captures' do
+    expect_offense(<<~RUBY)
+      /(?<foo>FOO)(?<bar>BAR)/ =~ "FOOBAR"
+      puts $3
+           ^^ Do not use out of range reference for the Regexp.
+    RUBY
+  end
+
+  it 'registers an offense when out of range references are used for numbered captures' do
+    expect_offense(<<~RUBY)
+      /(foo)(bar)/ =~ "foobar"
+      puts $3
+           ^^ Do not use out of range reference for the Regexp.
+    RUBY
+  end
+
+  it 'registers an offense when out of range references are used for mix of numbered and named captures' do
+    expect_offense(<<~RUBY)
+      /(?<foo>FOO)(BAR)/ =~ "FOOBAR"
+      puts $2
+           ^^ Do not use out of range reference for the Regexp.
+    RUBY
+  end
+
+  it 'registers an offense when out of range references are used for non captures' do
+    expect_offense(<<~RUBY)
+      /bar/ =~ 'foo'
+      puts $1
+           ^^ Do not use out of range reference for the Regexp.
+    RUBY
+  end
+
+  it 'does not register offense to a regexp with valid references for named captures' do
+    expect_no_offenses(<<~RUBY)
+      /(?<foo>FOO)(?<bar>BAR)/ =~ "FOOBAR"
+      puts $1
+      puts $2
+    RUBY
+  end
+
+  it 'does not register offense to a regexp with valid references for numbered captures' do
+    expect_no_offenses(<<~RUBY)
+      /(foo)(bar)/ =~ "foobar"
+      puts $1
+      puts $2
+    RUBY
+  end
+
+  it 'does not register offense to a regexp with valid references for a mix named and numbered captures' do
+    expect_no_offenses(<<~RUBY)
+      /(?<foo>FOO)(BAR)/ =~ "FOOBAR"
+      puts $1
+    RUBY
+  end
+
+  # See https://github.com/rubocop-hq/rubocop/issues/8083
+  it 'does not register offense when using a Regexp cannot be processed by regexp_parser gem' do
+    expect_no_offenses(<<~'RUBY')
+      /data = ({"words":.+}}}[^}]*})/m
+    RUBY
```

Comment:
Might as well add `$3` or something. Hopefully the Regexp parser will handle this one day and that test will fail, but right now it would always pass.

Reply:
There's still no reason to raise an offense here once parsing is fixed, right?

Reply:
I have added $1 in last commit to fix this. please refer [this](https://github.com/rubocop-hq/rubocop/pull/8407/commits/f97bd95cdde545bad681cdf63c8e19fc5c5fa1f9#diff-e36629e056bdba9159de0a2347dc3807R74)